File: 6.MinimumSkew/Minimumskew.cpp
//
// Created by shvet on 17/06/2024.
//
#include <string>
#include <iostream>
#include <fstream>
#include <vector>

void read_genome(const std::string& path, std::string& sequence){
    std::ifstream file(path);
    if (file.is_open()){
        std::cout << "File is opened!\n";
    }
    std::getline(file, sequence);
}

std::vector<int> MinimumSkew(const std::string& sequence){
    std::vector<int> skew = {0};
    std::vector<int> mn_indices;
    int mn = 0;
    for(size_t i = 1; i <= sequence.size(); i++){
        if (sequence[i - 1] == 'C'){
            skew.push_back(skew[i-1]-1);
        } else if (sequence[i - 1] == 'G'){
            skew.push_back(skew[i-1]+1);
        } else {
            skew.push_back(skew[i-1]);
        }
        if (skew.back() < mn){
            mn = skew.back();
            mn_indices = {static_cast<int>(i)};
        } else if (skew.back() == mn) {
            mn_indices.push_back({static_cast<int>(i)});
        }
    }
    return mn_indices;
}

void prt_vect(const std::vector<int>& vec){
    for(const auto& i : vec){
        std::cout << i << " ";
    }
    std::cout << "\n";
}

File: 6.MinimumSkew/Minimumskew.go
package main

import "fmt"

func minimum_skew(genome string) []int {
	skew := []int{0}
	var mn_indices []int
	mn := 0
	for i := 1; i < len(genome)+1; i++ {
		fmt.Println(i)
		if genome[i-1] == 'C' {
			skew = append(skew, skew[i-1]-1)
		} else if genome[i-1] == 'G' {
			skew = append(skew, skew[i-1]+1)
		} else {
			skew = append(skew, skew[i-1])
		}

		if skew[len(skew)-1] < mn {
			mn = skew[len(skew)-1]
			mn_indices = []int{i}
		} else if skew[len(skew)-1] == mn {
			mn_indices = append(mn_indices, i)
		}
	}
	return mn_indices
}

func main() {
	dna := "TAAAGACTGCCGAGAGGCCAACACGAGTGCTAGAACGAGGGGCGTAAACGCGGGTCCGAT"
	result := minimum_skew(dna)
	fmt.Println(result)
}
